lib/jxl/encode.cc

```diff
@@ -489,23 +489,34 @@ JxlEncoderStatus JxlEncoderSetBasicInfo(JxlEncoder* enc,
     return JXL_ENC_NOT_SUPPORTED;
   }
 
-  switch (info->alpha_bits) {
-    case 0:
-      break;
-    case 32:
-    case 16:
-      enc->metadata.m.SetAlphaBits(16);
-      break;
-    case 8:
-      enc->metadata.m.SetAlphaBits(info->alpha_bits);
-      break;
-    default:
-      return JXL_ENC_ERROR;
-  }
   // The number of extra channels includes the alpha channel, so for example and
   // RGBA with no other extra channels, has exactly num_extra_channels == 1
   enc->metadata.m.num_extra_channels = info->num_extra_channels;
   enc->metadata.m.extra_channel_info.resize(enc->metadata.m.num_extra_channels);
+  if (info->num_extra_channels == 0 && info->alpha_bits) {
+    return JXL_API_ERROR(
+        "when alpha_bits is non-zero, the number of channels must be at least "
+        "1");
+  }
+  // If the user provides non-zero alpha_bits, we make the channel info at index
+  // zero the appropriate alpha channel.
+  if (info->alpha_bits) {
+    JxlExtraChannelInfo channel_info;
+    JxlEncoderInitExtraChannelInfo(JXL_CHANNEL_ALPHA, &channel_info);
+    switch (info->alpha_bits) {
+      case 32:
+      case 16:
+        channel_info.bits_per_sample = 16;
+        break;
+      case 8:
+        channel_info.bits_per_sample = info->alpha_bits;
+        break;
+      default:
+        return JXL_ENC_ERROR;
+    }
+    JxlEncoderSetExtraChannelInfo(enc, 0, &channel_info);
```

Comment:
I suppose the aim is to avoid having to manually add the alpha channel since it's a common case, and to use e.g.  `basicinfo.alpha_bits = 8;` as a one-line way to add an alpha channel.

But if you want to, you could also do (in an application)
```
JxlExtraChannelInfo alpha;
JxlEncoderInitExtraChannelInfo(JXL_CHANNEL_ALPHA, &alpha);
alpha.bits_per_sample = 8;
JxlEncoderSetExtraChannelInfo(enc, 0, &alpha);
```

and that should also work, right? With or without setting `alpha_bits` in the basic info. Do I understand that correctly?

I'm not sure if I like having two ways to do the same thing - I think it would have been better if we didn't treat alpha as something special and have alpha_bits in the basic info. For passing pixel buffers it is convenient to have a way to interleave an extra channel with the RGB, but RGBA is not the only thing that is useful: CMYK (i.e. RGBK with an icc profile that makes RGB behave like CMY) and e.g. RGB+depth or RGB+thermal could also be useful. More generally it might also be useful to do more than 4 channels, e.g. CMYKA is a thing and I can also imagine e.g. RGB+depth+thermal to be useful, but if we don't want to go there, then "color + extra channel 0" would already be more useful than requiring that channel to represent alpha.

Reply:
> I suppose the aim is to avoid having to manually add the alpha channel since it's a common case, and to use e.g. `basicinfo.alpha_bits = 8;` as a one-line way to add an alpha channel.
> 
> But if you want to, you could also do (in an application)
> 
> ```
> JxlExtraChannelInfo alpha;
> JxlEncoderInitExtraChannelInfo(JXL_CHANNEL_ALPHA, &alpha);
> alpha.bits_per_sample = 8;
> JxlEncoderSetExtraChannelInfo(enc, 0, &alpha);
> ```
> 
> and that should also work, right? With or without setting `alpha_bits` in the basic info. Do I understand that correctly?
Yes, that should work. Also `bits_per_sample` will be initialized with `8` by `JxlEncoderInitExtraChannelInfo`, so no need for that line if you have 8 bits in the alpha channel. 
 Right now `JxlEncoderSetExtraChannelInfo` does not check for validity of the `bits_per_sample` value.

I guess the question is how much boiler-plate we expect from an user of the api.
> 
> I'm not sure if I like having two ways to do the same thing - I think it would have been better if we didn't treat alpha as something special and have alpha_bits in the basic info. For passing pixel buffers it is convenient to have a way to interleave an extra channel with the RGB, but RGBA is not the only thing that is useful: CMYK (i.e. RGBK with an icc profile that makes RGB behave like CMY) and e.g. RGB+depth or RGB+thermal could also be useful. More generally it might also be useful to do more than 4 channels, e.g. CMYKA is a thing and I can also imagine e.g. RGB+depth+thermal to be useful, but if we don't want to go there, then "color + extra channel 0" would already be more useful than requiring that channel to represent alpha.
Yeah, I think the "A"-case is much more common than the "K" or "depth" or "thermal" case. As of now, 
`JxlEncoderAddImageFrame` is hard-coded to search for the first alpha channel and put the 4th (or 2nd) interleaved channel there. I'm not sure if we really want to adapt it to allow putting the interleaved 4th (or 2nd) channel to a channel of a different kind.

Reply:
I think it's probably easier to make the interleaving/de-interleaving function use whatever extra channel 0 happens to be, instead of checking specifically for the alpha channel. I agree that A is way more common than anything else (but this could change over time as cameras with depth sensors become more common).

Anyway, that's all not really related to this PR, so I'll close this comment thread :)